osutil/disks/disks.go

```diff
@@ -0,0 +1,362 @@
+	out, err := udevadmProperties(device)
+	if err != nil {
+		return nil, osutil.OutputErr(out, err)
+	}
+	r := bytes.NewBuffer(out)
+
+	return parseUdevProperties(r)
+}
+
+func parseUdevProperties(r io.Reader) (map[string]string, error) {
+	m := make(map[string]string)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		strs := strings.SplitN(scanner.Text(), "=", 2)
+		if len(strs) != 2 {
+			// bad udev output?
+			continue
+		}
+		m[strs[0]] = strs[1]
+	}
+
+	return m, scanner.Err()
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point.
+func DiskFromMountPoint(mountpoint string, opts *Options) (Disk, error) {
+	// call the unexported version that may be mocked by tests
+	return diskFromMountPoint(mountpoint, opts)
+}
+
+type disk struct {
+	major int
+	minor int
+	// partitions is a map of label -> partition uuid for now
+	// eventually this may be expanded to be more generally useful
+	partitions map[string]string
+}
+
+// diskFromMountPointImpl returns a Disk for the underlying mount source of the
+// specified mount point. For mount points which have sources that are not
+// partitions, and thus are a part of a disk, the returned Disk refers to the
+// volume/disk of the mount point itself.
+func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	d := &disk{}
+	var mountpointSrc string
+	// loop over the mount entries in reverse order to prevent shadowing of a
+	// particular mount on top of another one
+	for i := len(mounts) - 1; i >= 0; i-- {
+		if mounts[i].MountDir == mountpoint {
+			d.major = mounts[i].DevMajor
+			d.minor = mounts[i].DevMinor
+			mountpointSrc = mounts[i].MountSource
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
+	}
+
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	props, err := udevProperties(mountpointSrc)
+	if err != nil && props == nil {
+		// only fail here if props is nil, if it's available we validate it
+		// below
+		return nil, fmt.Errorf("cannot find disk for partition %s: %v", mountpointSrc, err)
+	}
+
+	if opts != nil && opts.IsDecryptedDevice {
+		// verify that the mount point is indeed a mapper device, it should:
+		// 1. have DEVTYPE == disk from udev
+		// 2. have dm files in the sysfs entry for the maj:min of the device
+		if props["DEVTYPE"] != "disk" {
+			// not a decrypted device
+			return nil, fmt.Errorf("mountpoint source %s is not a decrypted device", mountpointSrc)
+		}
+
+		// TODO:UC20: these files should also be readable through udev env
+		//            properties, but currently aren't available for some reason
```

Comment:
so it seems I may have been wrong about this, as these udev env vars are indeed available when I just did a fresh re-build of the initrd to test something else. Unclear what changed, but perhaps my kernel I was testing with prior was too old... Anyways when I add this command to the-tool, I can now see the vars defined:

```
May 18 20:00:51 ubuntu the-tool[368]: + udevadm info --query property --name /dev/mapper/ubuntu-data-60ed8a96-0a05-426c-a188-92e114676670
May 18 20:00:51 ubuntu the-tool[562]: DEVPATH=/devices/virtual/block/dm-0
May 18 20:00:51 ubuntu the-tool[562]: DEVNAME=/dev/dm-0
May 18 20:00:51 ubuntu the-tool[562]: DEVTYPE=disk
May 18 20:00:51 ubuntu the-tool[562]: MAJOR=253
May 18 20:00:51 ubuntu the-tool[562]: MINOR=0
May 18 20:00:51 ubuntu the-tool[562]: SUBSYSTEM=block
May 18 20:00:51 ubuntu the-tool[562]: USEC_INITIALIZED=2588568
May 18 20:00:51 ubuntu the-tool[562]: DM_UDEV_DISABLE_LIBRARY_FALLBACK_FLAG=1
May 18 20:00:51 ubuntu the-tool[562]: DM_UDEV_PRIMARY_SOURCE_FLAG=1
May 18 20:00:51 ubuntu the-tool[562]: DM_UDEV_RULES=1
May 18 20:00:51 ubuntu the-tool[562]: DM_UDEV_RULES_VSN=2
May 18 20:00:51 ubuntu the-tool[562]: DM_NAME=ubuntu-data-60ed8a96-0a05-426c-a188-92e114676670
May 18 20:00:51 ubuntu the-tool[562]: DM_UUID=CRYPT-LUKS2-5bde46be89144808b02ee317de5ecc83-ubuntu-data-60ed8a96-0a05-426c-a188-92e114676670
May 18 20:00:51 ubuntu the-tool[562]: DM_SUSPENDED=0
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_LABEL=ubuntu-data
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_LABEL_ENC=ubuntu-data
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_UUID=8346dfbe-a1c2-48fc-bfbb-bb6d977c6855
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_UUID_ENC=8346dfbe-a1c2-48fc-bfbb-bb6d977c6855
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_VERSION=1.0
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_TYPE=ext4
May 18 20:00:51 ubuntu the-tool[562]: ID_FS_USAGE=filesystem
May 18 20:00:51 ubuntu the-tool[562]: DEVLINKS=/dev/disk/by-id/dm-uuid-CRYPT-LUKS2-5bde46be89144808b02ee317de5ecc83-ubuntu-data-60ed8a96-0a05-426c-a188-92e114676670 /dev/disk/by-uuid/8346dfbe-a1c2-48fc-bfbb-bb6>
May 18 20:00:51 ubuntu the-tool[562]: TAGS=:systemd:
```

however, running that same command in userspace does not show these properties:

```
$ udevadm info --query property --name /dev/mapper/ubuntu-data-60ed8a96-0a05-426c-a188-92e114676670
DEVPATH=/devices/virtual/block/dm-0
DEVNAME=/dev/dm-0
DEVTYPE=disk
MAJOR=253
MINOR=0
SUBSYSTEM=block
USEC_INITIALIZED=4662018
DM_UDEV_DISABLE_SUBSYSTEM_RULES_FLAG=1
DM_UDEV_DISABLE_DISK_RULES_FLAG=1
DM_UDEV_DISABLE_OTHER_RULES_FLAG=1
SYSTEMD_READY=0
TAGS=:systemd:
```

So I'm at a bit of a loss why it doesn't work from userspace, but I think we can probably switch to using only udevadm if we wanted to since it works from the initrd now, though maybe it's just racy

Reply:
maybe something related to what we do to keep things mounted across the pivot?

Reply:
could be, I think I have an idea I need to explore tomorrow, but for now let's keep with the sysfs parsing before we switch over to udev in the initrd.